data-pipeline-ffi/src/error.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use data_pipeline::trace_exporter::error::{
+    BuilderErrorKind, NetworkErrorKind, TraceExporterError,
+};
+use std::ffi::{c_char, CString};
+use std::fmt::Display;
+use std::io::ErrorKind as IoErrorKind;
+
+/// Represent error codes that `Error` struct can hold
+#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ExporterErrorCode {
+    AddressInUse,
+    ConnectionAborted,
+    ConnectionRefused,
+    ConnectionReset,
+    HttpBodyFormat,
+    HttpBodyTooLong,
+    HttpClient,
+    HttpParse,
+    HttpServer,
+    HttpWrongStatus,
+    InvalidArgument,
+    InvalidData,
+    InvalidInput,
+    InvalidUrl,
+    IoError,
+    NetworkUnknown,
+    Serde,
+    TimedOut,
+}
+
+impl Display for ExporterErrorCode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::AddressInUse => write!(f, "Address already in use"),
+            Self::ConnectionAborted => write!(f, "Connection aborted"),
```

Comment:
/question

do you think it would make sense to return the actual message, when the root cause of the error is not us? This would help in debugging and reduce the number of mapping we own.

Those error could be even more informative.

Reply:
We're actually sending the proper message when the error is originated internally. In the `from` trait [implementation](https://github.com/DataDog/libdatadog/pull/768/files#diff-46620090d8d466154fc006ef9bfff44f8144f70c353356664e866b817790f179R77-R117) you'll see that we're translating to string the `TraceExporterError` contents which means, in the case of Io and Hyper errors, we'll use their own [display representation](https://github.com/DataDog/libdatadog/pull/768/files#diff-1d490aa446c998ae0124ca12a84ec4a2094fc80f9714bb705351599399d63cd4R58-R62).
This trait implementation is only used when the error is caused in the ffi layer as [in parsing arguments](https://github.com/DataDog/libdatadog/pull/768/files#diff-6b9976c838fe0d73c7e4b54e1bb1c3dbee58eb65441f46ff61b6ca8364d468c1R267-R268)

Reply:
Perfect